HealthMonitoring.AcceptanceTests/Helpers/Selenium/WebDriverExtensions.cs

```diff
@@ -2,78 +2,147 @@
+                $"Page with url {expectedUrl} did not loaded!");
+        }
 
-        public static void LoadUrl(this IWebDriver driver, string url)
+
+        public static IWebElement ExplicitWaitElementIsRendered(this IWebDriver driver, Func<IWebDriver, IWebElement> selector)
         {
-            driver.Navigate().GoToUrl(url);
+            return driver.ExplicitWait(
+                TimeSpan.FromSeconds(60),
+                selector,
+                $"Element with selector:{selector} could not be rendered");
         }
 
-        public static string WaitUntilPageIsChanged(this IWebDriver driver, string expectedUrl)
+        public static string ExplicitWaitTextIsRendered(this IWebDriver driver, Func<IWebDriver, IWebElement> selector)
         {
-            return Wait.Until(
+            return driver.ExplicitWait(
                 Timeouts.Default,
-                () => driver.Url,
-                url => string.Equals(url, expectedUrl, StringComparison.CurrentCultureIgnoreCase),
-                $"Page with url {expectedUrl} did not loaded!");
+                selector,
+                $"Element with selector:{selector} could not be rendered").Text;
         }
 
-        public static void RetryTimeout(this IWebDriver driver, TimeSpan period)
+        public static IEnumerable<IWebElement> ExplicitWaitElementsAreRendered(this IWebDriver driver, 
+            Func<IWebDriver, IEnumerable<IWebElement>> selector)
         {
-            driver.Manage().Timeouts().ImplicitWait = period;
+             return driver.ExplicitWait(
+                 Timeouts.Default,
+                 selector,
+                 $"Elements with selector:{selector} could not be rendered");
         }
 
-        private static IWebElement FindElement(IWebDriver driver, By selector)
+        public static bool ExplicitWaitUntilPageIsChanged(this IWebDriver driver, Func<IWebDriver, bool> selector, string expectedUrl)
         {
+            return driver.ExplicitWait(
+                Timeouts.Default,
+                selector,
+                $"Page with url {expectedUrl} did not loaded!");
+        }
+
+        public static T ImplicitWait<T, TWebDriver>(this TWebDriver webDriver,
+            TimeSpan implicitWait,
+            Func<TWebDriver, T> selector,
+            Func<T, bool> predicate,
+            string errorMessage)
+            where TWebDriver : IWebDriver
+        {
+            T valueFound = default(T);
+
             try
             {
-                return driver.FindElement(selector);
+                webDriver.Manage().Timeouts().ImplicitWait = implicitWait;
+
+                valueFound = selector(webDriver);
+
+                if (predicate(valueFound))
+                    return valueFound;
+
+                throw new NoSuchElementException($"{valueFound} not found");
             }
-            catch (NoSuchElementException)
+            catch (NoSuchElementException e)
```

Comment:
Why do we need here this try catch block if we rethrowing another exception?
We can remove try catch and just throw new TimeoutException
in this case we can remove  this line T valueFound = default(T);